### PR TITLE
Refactor read parquet table to reduce file list scan based on available partitions

### PR DESCRIPTION
**Pull request by murgua**
Refactor read parquet table to reduce file list scan based on available partitions

*Issue #, if available:*

*Description of changes:*
"Refactor read parquet table to reduce file list scan based on available partitions to improve response times and memory usage

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 4fb6c3c5fd571504c261539b1ae0af38f235d530
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f53a911c-9274-4c9c-8bc2-a6d9031fbf9a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 57aaf05ab4ef37febc34580a74df8d48c3e9f376
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=80490042-d44c-4a3e-bffa-9d98c158709d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: e295af3eaa38988211114d21208a034b011d22b0
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8fcc8de8-8573-45ff-9afb-86a170be1884%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 252393af800b2e888c97e5551e16eab1b7024d28
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=73fa25b5-cb03-4ce1-ab71-d6a5c2999ccc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_read_parquet.py [hunk 1]
```diff
@@ -524,6 +525,7 @@ def read_parquet(
     last_modified_end: Optional[datetime.datetime] = None,
     boto3_session: Optional[boto3.Session] = None,
     s3_additional_kwargs: Optional[Dict[str, Any]] = None,
+    path_root: Optional[str] = None,
```
Comment (kukushking) [inline awswrangler/s3/_read_parquet.py:528]: Would be great to have this parameter explained in the docstrings

**Review by kukushking — APPROVED** [review]

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 18fcb5c69a597ff5723bc474718bc5353865680f
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4d83a1b5-82a3-4b1a-8f6a-1c892e74d5b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 98176dfc3bdc1f904bf9c0c69bd33584022c577f
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d48ac489-2115-450a-afeb-c73d0d384fb5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_read_parquet.py [hunk 2]
```diff
@@ -867,11 +873,37 @@ def read_parquet_table(
...
+    # Then list objects & process individual object keys under path_root
+    if partition_filter is not None:
+        available_partitions_dict = _get_partitions(
+            database=database,
+            table=table,
+            catalog_id=catalog_id,
+            boto3_session=boto3_session,
+        )
+        available_partitions = list(available_partitions_dict.keys())
+        if len(available_partitions) > 0:
```
Comment (jaidisido) [inline awswrangler/s3/_read_parquet.py:888]: Nitpick but would `if available_partitions` do it?

File: tests/test_s3_parquet.py [hunk 3]
```diff
@@ -70,6 +70,27 @@ def test_read_parquet_filter_partitions(path, use_threads):
     assert df2.c2.astype(int).sum() == 0
 
 
+def test_read_parquet_table(path, glue_database, glue_table):
+    df = pd.DataFrame({"c0": [0, 1, 2], "c1": [0, 1, 2], "c2": [0, 0, 1]})
+    wr.s3.to_parquet(df, path, dataset=True, database=glue_database, table=glue_table)
+    df_out = wr.s3.read_parquet_table(table=glue_table, database=glue_database)
+    assert df_out.shape == (3, 3)
```
Comment (jaidisido) [inline tests/test_s3_parquet.py:77]: What is the point of this test? Was it just that we did not have one for plain `read_parquet_table`?

File: awswrangler/s3/_read_parquet.py [hunk 4]
```diff
@@ -867,11 +873,37 @@ def read_parquet_table(
...
+            paths = []
+            path_root = path
+            partitions: Union[str, List[str]] = _apply_partition_filter(
+                path_root=path, paths=available_partitions, filter_func=partition_filter
```
Comment (jaidisido) [inline awswrangler/s3/_read_parquet.py:892]: Should it not be `path_root=path_root`? Realize it's equivalent but just for consistency

File: awswrangler/s3/_read_parquet.py [hunk 5]
```diff
@@ -506,6 +507,7 @@ def _read_parquet(
 
 def read_parquet(
```
Comment (jaidisido) [inline awswrangler/s3/_read_parquet.py:508]: Should we be doing this for `read_text` methods (e.g. `read_csv`) as well?

File: awswrangler/s3/_read_parquet.py (hunk 2 above)
Comment (kukushking) [reply awswrangler/s3/_read_parquet.py:888]: Empty list is falsey so yeah that would do. I'll change.

File: tests/test_s3_parquet.py (hunk 3 above)
Comment (kukushking) [reply tests/test_s3_parquet.py:77]: Yes, that's just a test case for basic scenario.

File: awswrangler/s3/_read_parquet.py (hunk 4 above)
Comment (kukushking) [reply awswrangler/s3/_read_parquet.py:892]: Yeah, that'll do, too.

File: awswrangler/s3/_read_parquet.py (hunk 5 above)
Comment (kukushking) [reply awswrangler/s3/_read_parquet.py:508]: Do you mean the same partition optimisation? We're missing a few key methods for others at this point like `read_csv_table` (that's when `path_root` really becomes useful), but ideally yes.

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: c1f545c1bdd7c9ca4beb722a4024dfbfa823657b
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=85585809-1e9d-4beb-8c01-fb18951c6527%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: a07ebb8150ae9b4d8f8fa43a9b88555ba63917db
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=98ad62d0-c2b8-4672-a834-250225b75453%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**